src/java.base/share/classes/java/util/zip/ZipInputStream.java

```diff
@@ -627,6 +640,91 @@ private void readEnd(ZipEntry e) throws IOException {
+     * as 64-bit numbers instead of the regular 32-bit numbers.
+     *
+     * Returns true if the LOC has the 'streaming mode' flag set, the 'compressed size' and
+     * 'uncompressed size' are set to the Zip64 magic value 0xFFFFFFFF, and the LOC's extra field
+     * contains a Zip64 field where both 'Original Size' and 'Compressed Size' are set to zero.
+     *
+     * @param extra the LOC extra field to look for a Zip64 field in
+     * @param flag the value of the 'general purpose bit flag' field in the LOC
+     * @param csize the value of the 'compressed size' field in the LOC
+     * @param size  the value of the 'uncompressed size' field in the LOC
+     */
+    private boolean expect64BitDataDescriptor(byte[] extra, int flag, long csize, long size)  {
+        // The LOC's 'general purpose bit flag' 3 must indicate use of a Data Descriptor
+        if ((flag & 8) == 0) {
+            return false;
+        }
+
+        // The LOC's 'compressed size' and 'uncompressed size' must both be marked for Zip64
+        if (csize != ZIP64_MAGICVAL || size != ZIP64_MAGICVAL) {
+            return false;
+        }
+
+        // Look for a Zip64 field valid for use with data descriptors
+        int headerSize = 2 * Short.BYTES; // id + size
+        if (extra != null) {
+            for (int i = 0; i + headerSize < extra.length;) {
+                int id = get16(extra, i);
+                int dsize = get16(extra, i + Short.BYTES);
+                if (i + headerSize + dsize > extra.length) {
+                    return false; // Invalid size
+                }
+                if (isZip64DataDescriptorField(id, extra, i + headerSize, dsize)) {
+                    return true;
+                }
+                i += headerSize + dsize;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Returns true if the given extra field is a Zip64 field with both 'Original Size'
+     * and 'Compressed Size' fields set to zero, making it valid for use with a Data Descriptor.
+     * <p>
+     * The order of the Zip64 fields is fixed, but the fields MUST
+     * only appear if the corresponding LOC field is set to 0xFFFFFFFF:
+     * Uncompressed Size - 8 bytes
+     * Compressed Size   - 8 bytes
+     * <p>
+     * This entry in the Local header MUST include BOTH original
+     * and compressed file size fields.
+     * <p>
+     * See PKWare APP.Note Section 4.5.3 for more details
+     *
+     * @param headerId the extra field header id
+     * @param extra the array where the extra block is stored
+     * @param blockStart the offset into the array where the extra block starts
+     * @param blockSize the size of the extra field data block
+     * @return true if the extra field is a Zip64 extra field compatible with data descriptors
+     */
+    private static boolean isZip64DataDescriptorField(int headerId, byte[] extra, int blockStart, int blockSize) {
```

Comment:
I understand the goals of this method - what it's trying to do is, assure the caller that the extra field/block actually is a zip64 extra block. That assurance is then used to access the data descriptor content as 8 byte fields. 
However, I think in this proposed implementation of this method we are perhaps doing a bit too much. Specifically, I don't think we should check what values have been stamped for "Original size" and "Compressed size" fields of this zip64 block. I think, those values (presence or absence) shouldn't play a role in deciding whether we have to read a data descriptor size fields as 8 bytes. Doing these checks for these zip64 original/compressed size fields, I think will open up more permutations about which zip entries get processed as 8 byte data descriptors.

Given the context in which this method is used, I think the only checks that we should do in this method is to verify that the header id is `ZIP64_EXTID`. 
Perhaps then this `isZip64DataDescriptorField(...)` won't be needed and we can just inline that `headerid == ZIP64_EXTID` check inline in the implementation of `expect64BitDataDescriptor(...)` method